bftengine/src/bftengine/Crypto.cpp

```diff
@@ -58,23 +58,19 @@ namespace impl {
 //#define RSA_STANDARD PKCS1v15
 //#define RSA_STANDARD OAEP<SHA>
 
-static RandomPool sGlobalRandGen;  // not thread-safe !!
-
-void CryptographyWrapper::init(const char* randomSeed) {
-  string s(randomSeed);
-  if (s.length() < 16) s.resize(16, ' ');
-  sGlobalRandGen.IncorporateEntropy((CryptoPP::byte*)s.c_str(), s.length());
-
-  VERIFY(DigestUtil::digestLength() == DIGEST_SIZE);
-
-  // Initialize RELIC library for BLS threshold signatures
-  BLS::Relic::Library::Get();
-}
+// Wrap RandomPool in a struct which initialises it. The wrapper is used as static thread local variable so that each
+// thread has got its own instance. The reason for this complication is thread safety. This RandomPool is used by the
+// SigManager which is a singleton used by many threads concurrently.
+struct RandomPoolWrapper : RandomPool {
+  RandomPoolWrapper() {
+    SecByteBlock seed(32);
+    // this might block for a while - 'man 4 random' for details
+    OS_GenerateRandomBlock(true, seed, seed.size());
```

Comment:
You don't need to block here and read from `/dev/urandom` instead. All modern systems have a HWRNG.

Reply:
Doesn't /dev/random use the same HWRNG? If yes - I think it's safer this way.

Reply:
Why is it safer? (It is a loaded question, no need to answer) :)
https://www.2uo.de/myths-about-urandom/#experts

Reply:
We can talk more offline about /dev/random vs /dev/urandom. Either is fine if blocking doesn't become an issue in this case.

Reply:
Great read, @robem . And thanks for the feedback. I'll use the non-blocking version (/dev/urandom).